internal/planner/planner.go

```diff
@@ -1580,10 +1579,16 @@ func (p *Planner) planRefData(virtual *ruletrie, base *baseptr, ref ast.Ref, ind
 		}); err != nil {
 			return err
 		}
+
+		p.ltarget = base.local
+		return p.planRefDataBaseScan(base.path, ref, index, lexclude, iter)
 	}
 
 	p.ltarget = base.local
-	return p.planRefDataBaseScan(base.path, ref, index, lexclude, iter)
+	base.path = append(base.path, ref[index])
+	return p.planRefRec(base.path, 0, func() error {
```

Comment:
It took me a little while to figure out what's happening here.

This code path is almost identical to the one right above except L1584 always generates a scan whereas L1589 won't generate a scan if the term (`ref[index]`) is ground or seen.

I was wondering if we could combine the code paths to simplify things but the `planRefRec` iterator doesn't provide the local variable for the key (which is required to check the exclude set).

I will say that I wonder if it would be better to just inline `planRefDataBaseScan` above as it's only called from one place and it's not doing very much (this was part of the confusion for me...I assumed more was going on when there wasn't)

Reply:
I've inlined it again 😃